---
title: "whole_pipeline"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## After preprocessing, there should be a series of input files:
1. X_nreads_processed.csv: number of reads mapping to each gene in the shape of (gene_symbol * sample_name), with experiment replicates merged together.
2. X_pheno.csv: the annotation for each cell/sample, one columns with sample name as rownames.

```{r load data}
#prefix<-"iPSC/all_"
#prefix<-"HSC/hHSC_"
#prefix<-"mouseSC/mouseSC_"
prefix<-"mouseSC2/mouseSC2_"
#species<-"human" 
species<-"mice"

filename<-paste(prefix, "nreads_processed.csv", sep="")
nreads= read.csv(filename,sep=" ",header = TRUE,stringsAsFactors = FALSE) 
gene.names<-rownames(nreads)

filename<-paste(prefix, "pheno.csv", sep="")
pheno= read.csv(filename,sep=" ",header = TRUE,stringsAsFactors = FALSE) 

pheno<-pheno[colnames(nreads),]
```

## Adding feature-level annotation (separately for human and mice) 

```{r adding gene features}
library(BiocFileCache)
library(biomaRt)
bmcache <- BiocFileCache("biomart", ask = FALSE)

if (species=="human"){
    ensembl <- useMart(biomart='ENSEMBL_MART_ENSEMBL', dataset="hsapiens_gene_ensembl", 
        host="aug2017.archive.ensembl.org") # Ensembl version 90.
    ensemblGenes <- getBM(attributes=c('ensembl_gene_id',  'chromosome_name', 'gene_biotype', 
        'external_gene_name', 'entrezgene'), filters="external_gene_name", 
        values=gene.names, mart=ensembl) 
    saveRDS(ensemblGenes, file=bfcnew(bmcache, "hg38.ensGene"))
    
    features <- ensemblGenes[match(gene.names, ensemblGenes$external_gene_name),]
    features$entrezgene <- gsub(" ", "", as.character(features$entrezgene))
    row.names(features) <- gene.names
    head(features)
}else{
  ensembl <- useMart(biomart='ensembl', dataset="mmusculus_gene_ensembl")
  ensemblGenes <- getBM(attributes=c('ensembl_gene_id',  'chromosome_name', 
                                     'external_gene_name'), filters="external_gene_name", 
                        values=gene.names, mart=ensembl) 
  
  human = useMart("ensembl",dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl",dataset = "mmusculus_gene_ensembl")
  homo_genes=getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol",
                    values = ensemblGenes$external_gene_name, mart=mouse,attributesL = c("hgnc_symbol","ensembl_gene_id"),
                    martL = human,
                    uniqueRows = T)
  homo_genes_dedup<- homo_genes %>% distinct(MGI.symbol, .keep_all = TRUE)
                                             
  features <- ensemblGenes[match(gene.names, ensemblGenes$external_gene_name),]
  row.names(features) <- gene.names
  features<-merge(features,homo_genes_dedup, by.x="external_gene_name",by.y="MGI.symbol", all=TRUE)
    
}


```

## Mitochondrial genes and spike-in transcripts are defined.

```{r defining spike in and mitochondrial genes}
is.spike <- grepl("^ERCC-", gene.names)
summary(is.spike)
is.mito <- !is.na(features[!is.spike,]$chromosome_name) & features[!is.spike,]$chromosome_name=="MT"
summary(is.mito)
#先除spike,再除mito
```

## Constructing the `SingleCellExperiment` object.

```{r construct SCE}
library(SingleCellExperiment)
if (nrow(nreads) == length(is.spike)){
  sce <- SingleCellExperiment(list(counts=as.matrix(nreads[!is.spike,])), 
    rowData=features[!is.spike,], colData=pheno)
}
if (sum(is.spike)!=0){
  altExp(sce,"ERCC")<-SingleCellExperiment(list(counts=as.matrix(nreads[is.spike,])     ), rowData=features[is.spike,], colData=pheno)
}

sce <- splitAltExps(sce, ifelse(is.mito, "mito", "gene"))
sce

```

## Do QC: get rid of single cells with low library size, low number of features detected, high proportion of mitochondria and spike-in reads.

```{r QC highEx, fig.height=3, fig.width=9}
sce_qc <- perCellQCMetrics(sce) 
colData(sce)<-cbind(colData(sce), sce_qc)
colnames(colData(sce))
colData(sce)$cell_type<-factor(colData(sce)$cell_type)

colData(sce)$discard=discard
if (sum(is.spike)!=0){
  multiplot(cols=2,
    plotColData(sce, x="cell_type", y="sum",colour_by ="discard") + scale_y_log10() + ggtitle("Total count"),   # sum of counts, or library size
    plotColData(sce, x="cell_type", y="detected",colour_by ="discard")+ ggtitle("Detected features"),  # number of detected features
    plotColData(sce, x="cell_type", y="altexps_ERCC_percent",colour_by ="discard")+ ggtitle("Spike-in percent"),
    plotColData(sce, x="cell_type", y="altexps_mito_percent",colour_by ="discard")+ ggtitle("Mito percent")
)
}else{
    multiplot(cols=3,
    plotColData(sce, x="cell_type", y="sum",colour_by ="discard") + scale_y_log10() + ggtitle("Total count"),   # sum of counts, or library size; colour_by ="discard"
    plotColData(sce, x="cell_type", y="detected",colour_by ="discard") + ggtitle("Detected features"),  # number of detected features
    plotColData(sce, x="cell_type", y="altexps_mito_percent",colour_by ="discard") + ggtitle("Mito percent")
)
}

#mouseSC2
#multiplot(cols=3,
#    plotColData(sce,  y="sum",colour_by ="discard") + scale_y_log10() + ggtitle("Total #count"),   # sum of counts, or library size; colour_by ="discard"
#    plotColData(sce,  y="detected",colour_by ="discard") + ggtitle("Detected features"), # # number of detected features
#    plotColData(sce, y="altexps_mito_percent",colour_by ="discard") + ggtitle("Mito #percent")
#)

```
 
 
## remove cells based on some criteria (for each batch or each cell type)
 
```{r remove cells}
libsize.drop <- isOutlier(sce$sum, nmads=3, type="lower", log=TRUE, batch=sce$cell_type) 
feature.drop <- isOutlier(sce$detected, nmads=3, type="lower", log=TRUE, batch=sce$cell_type) 

# Or use fixed threshold: feature.drop <- sce$detected < 5000

mito.drop <- isOutlier(sce$altexps_mito_percent, nmads=3, type="higher", batch=sce$cell_type)
if (sum(is.spike)!=0){
  spike.drop <- isOutlier(sce$altexps_ERCC_percent, nmads=3, type="higher", batch=sce$cell_type)
  discard <- libsize.drop | feature.drop | spike.drop | mito.drop
}else{
  discard <- libsize.drop | feature.drop | mito.drop
}

discard <- libsize.drop | feature.drop | mito.drop
data.frame(ByLibSize=sum(libsize.drop), ByFeature=sum(feature.drop), ByMito=sum(mito.drop),
           BySpike=sum(spike.drop), Remaining.in.batch=sum(!discard))
```

```{r}
plotColData(sce, x="sum", y="altexps_mito_percent", colour_by = "discard")
```



## Examining the genes after removing cells

```{r}
sce <- sce[, !discard]
dim(sce) 

ave.counts <- apply(counts(sce),1,mean)
rowData(sce)$AveCount <- ave.counts
hist(log10(ave.counts), breaks=100, main="", col="grey", xlab=expression(Log[10]~"average count"))

```
## the number of cells expressing each gene

```{r}
ncells <- scater::nexprs(sce, byrow=TRUE)
rowData(sce)$Ncells <- ncells
plot(ave.counts, ncells, xlab="Average count", ylab="Number of cells", log="x", pch=16, cex=0.5)
```


## most highly expressed genes
```{r highEx, fig.height=12, fig.width=6}
fontsize <- theme(axis.text=element_text(size=12), axis.title=element_text(size=16))
scater::plotHighestExprs(sce, n=50) + fontsize
```

## discard genes not expressed

```{r}
keep <- ave.counts > 0
sce <- sce[keep,]
summary(keep)
```

# Normalization of various technical biases

```{r}
set.seed(10000)
clusters <- quickCluster(sce, method="igraph", min.mean=1)
sce <- computeSumFactors(sce, cluster=clusters, min.mean=1)
summary(sizeFactors(sce))

plot(sizeFactors(sce), sce$sum/1e6, log="xy", ylab="Library size (millions)", 
    xlab="Size factor")
sce <- logNormCounts(sce,log = TRUE)
```

## clustering using two different approaches

## scran 在PCA之前没做scaling?

```{r SC3 clustering and seurat clustering highEx, fig.height=4, fig.width=6}

sce<- denoisePCA(sce, var.out, assay.type="logcounts")
ncol(reducedDim(sce))
pc.out <- attr(reducedDim(sce), "percentVar")
plot(pc.out)

# SC3 clustering
rowData(sce)$feature_symbol<-rowData(sce)$external_gene_name
sce<-sc3(sce,ks=4)

markers<-findMarkers(sce, groups=colData(sce)$sc3_4_clusters, pval.type="all", direction="up")
# all means the marker has to be highly expressed compared to all other groups, and up means only consider upregulated genes

# Seurat clustering by making a Seurat process
sce.seurat <- as.Seurat(sce, counts = "counts", data = "logcounts")
sce.seurat <- FindVariableFeatures(sce.seurat, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(sce.seurat)
sce.seurat <- ScaleData(sce.seurat, features = all.genes)#vars.to.regress = "percent.mt")  
sce.seurat <- RunPCA(sce.seurat, features = VariableFeatures(object = sce.seurat))

sce.seurat <- FindNeighbors(sce.seurat, dims = 1:10)
sce.seurat <- FindClusters(sce.seurat, resolution = 0.5)
head(Idents(sce.seurat), 5)  

colData(sce)$seurat_cluster<-Idents(sce.seurat)

set.seed(100)
sce <- runTSNE(sce, use_dimred="PCA", perplexity=30)
#tsne1 <- scater::plotTSNE(sce, colour_by="cell_type") +  fontsize #+ scale_fill_manual(breaks = c(2, 1),
                      # values=c("red", "blue"))+ labs(fill = "Cell type")


#pca1 <- plotPCA(sce, colour_by="cell_type") + fontsize #+ scale_fill_manual(breaks = c(2, 1),
                       #values=c("red", "blue")) + labs(fill = "Cell type")

tsne2 <- scater::plotTSNE(sce, colour_by="sc3_4_clusters") + fontsize + scale_fill_manual(breaks = c("1", "4", "3","2"),
                       values=c("red", "blue", "green","yellow"))+ labs(fill = "SC3 cluster")
pca2 <- plotPCA(sce, colour_by="sc3_4_clusters") + fontsize + scale_fill_manual(breaks = c("1", "4","3", "2"),
                       values=c("red", "blue", "green","yellow"))+ labs(fill = "SC3 cluster")
tsne3 <- scater::plotTSNE(sce, colour_by="hc_corelation_4") + fontsize + scale_fill_manual(breaks = c("1", "3", "2","4"),
                       values=c("red", "blue", "green","yellow"))+ labs(fill = "hierarchical cluster")
pca3 <- plotPCA(sce, colour_by="hc_corelation_4") + fontsize + scale_fill_manual(breaks = c( "1", "3","2","4"),
                       values=c("red", "blue", "green","yellow"))+ labs(fill = "hierarchical cluster")
multiplot(tsne2,tsne3 , pca2,pca3, cols=2)


```
## dataset1 specific: remove cells wrongly clustered by Seurat

```{r}
levels(colData(sce)$sc3_3_clusters)<-c("iPSC","NPC","MN")
is.outlier<- colData(sce)$sc3_3_clusters!=colData(sce)$cell_type
sce<-sce[,!is.outlier]
```
## dataset 4 specific:
```{r}
sample_cor <- cor( logcounts(sce)) 

#Transform the scale from correlations
sample_cor <- (1 - sample_cor) / 2

#Convert it to a distance object
d2 <- as.dist(sample_cor)

h_correlation <- hclust(d2, method="ward.D2")

colData(sce)$hc_corelation_4 <- factor(cutree(h_correlation,k = 4))
markers<-findMarkers(sce, groups=colData(sce)$hc_corelation_4, pval.type="all", direction="up")
```



```{r save the sce object}
filename<-paste(prefix, "sce.rds", sep="")
saveRDS(sce, file=filename)


```

